### DISCUSSION TITLE
NER, EntityRuler and Knowledge Base

### DISCUSSION THREAD

**Discussion by namesparse** [explosion/spaCy#8085] [Help: Coding & Implementations] [answered]
Firstly just to say that I love Spacy and am in awe of how much effort has gone into maintaining it. 

I am trying to understand the Knowledge Base and EntityRuler a little better so was looking for some idiot-proof explanations on how they could be used in real world. 

My use case is that I have a list of company names with an identifier and I’d like to know if the named entity extracted from a document has an identifier (from my list). It could be that the name is abbreviated / spelled incorrectly too. 

I know I could extract any named entities from the document then separately run some kind of similarity script against each one but I also want to then use the EntityLinker too so wanted to check if Spacy already did what I was looking for.

thanks a lot

**Comment by polm** [comment, marked as answer]
If you want to add IDs to mentions there is a feature in the EntityRuler for that, the docs are [here](https://spacy.io/usage/rule-based-matching#entityruler-ent-ids). 

The EntityLinker disambiguates mentions, it doesn't handle mention resolution. You can learn more about how it would be used at [the example project here](https://github.com/explosion/projects/tree/v3/tutorials/nel_emerson), but an example is that given a sentence like "Emerson won the game today" it would decide which specific person "Emerson" referred to. If you spelled it as "Emersen" it would not help you resolve that. So if all you have for companies is names it might not be what you need. 

However, this issue has been brought up with regards to the EntityLinker before, and you can define a custom `get_candidates` function that uses fuzzy matching to supply potential candidates. See https://github.com/explosion/spaCy/issues/4981. 

The actual fuzzy matching implementation is up to you. For random mispellings you'll probably want to look into Levenshtein distance or other string similarity metrics.

**Reply by namesparse** [reply]
Thanks very much for the detail - I’ll go through the links you kindly provided.
